File: services/gateway/routes/vc_api.py
"""Authenticated W3C VC-API compatibility boundary.

The endpoints adapt VC-API request and response shapes to Marty's ordinary
issuance and presentation-policy APIs. They do not select custody providers,
resolve private key references, or perform an independent verification.
Every request is authenticated and organization-authorized by the gateway
before it reaches this router.
"""

from __future__ import annotations

import json
import os
from typing import Any
from urllib.parse import unquote_to_bytes

from fastapi import APIRouter, HTTPException, Query, Request
from fastapi.responses import JSONResponse, Response
from gateway.models import IssuanceCreate
from gateway.proxy import get_http_client, get_registry, proxy_request
from gateway.routes.issuance import create_issuance
from pydantic import BaseModel, Field, ValidationError

router = APIRouter(prefix="/v1/vc-api", tags=["W3C VC API"])


class VerifyCredentialRequest(BaseModel):
    verifiableCredential: str | dict[str, Any]
    options: dict[str, Any] = Field(default_factory=dict)


class VerifyPresentationRequest(BaseModel):
    verifiablePresentation: str | dict[str, Any]
    options: dict[str, Any] = Field(default_factory=dict)


class IssueCredentialRequest(BaseModel):
    credential: dict[str, Any]
    options: dict[str, Any] = Field(default_factory=dict)


def _require_organization_authorization(
    request: Request,
    organization_id: str,
) -> None:
    """Defend the router boundary even if its middleware wiring regresses."""
    authorized_organization = getattr(request.state, "organization_id", None)
    if authorized_organization != organization_id:
        raise HTTPException(
            status_code=403,
            detail="Request is not authorized for this organization",
        )


def _credential_issuer_id(credential: dict[str, Any]) -> str | None:
    issuer = credential.get("issuer")
    if isinstance(issuer, str):
        return issuer
    if isinstance(issuer, dict) and isinstance(issuer.get("id"), str):
        return issuer["id"]
    return None


def _create_oid4vci_proof(issuer_url: str, nonce: str) -> str:
    """Generate a real, one-off holder proof using the released Rust binding."""
    try:
        from marty_rs import _marty_rs as binding
    except Exception as exc:  # pragma: no cover - release image invariant
        raise HTTPException(
            status_code=503, detail="Marty OID4VCI proof binding is unavailable"
        ) from exc
    try:
        return str(binding.oid4vci_create_proof_jwt(issuer_url, nonce))
    except Exception as exc:
        raise HTTPException(
            status_code=503, detail="could not generate OID4VCI holder proof"
        ) from exc


def _token_or_unsupported(
    value: str | dict[str, Any], field: str
) -> str | dict[str, Any]:
    if isinstance(value, str) and value.strip():
        return value
    if isinstance(value, dict):
        proof = value.get("proof")
        proofs = proof if isinstance(proof, list) else [proof]
        if any(
            isinstance(item, dict) and item.get("type") == "DataIntegrityProof"
            for item in proofs
        ):
            # Keep the document structured. The ordinary presentation-policy
            # service sends it to the released Rust verifier and fails closed.
            return value
        return _extract_jose_envelope(value, field)
    raise HTTPException(
        status_code=422,
        detail={
            "error": "unsupported_serialization",
            "error_description": (
                f"{field} must use a supported JWT, SD-JWT, mdoc, "
                "or VCDM Data Integrity serialization"
            ),
        },
    )


def _extract_jose_envelope(value: dict[str, Any], field: str) -> str:
    """Extract a JWT from the VCDM v2 JOSE envelope representation.

    This is deliberately a representation adapter only.  The extracted token
    is still sent to the ordinary Marty evaluator; the adapter neither trusts
    nor verifies the JWS itself.  JSON-LD Data Integrity objects remain
    explicitly unsupported.
    """
    context = value.get("@context")
    contexts = context if isinstance(context, list) else [context]
    types = value.get("type")
    identifier = value.get("id")
    expected_type = (
        "EnvelopedVerifiableCredential"
        if field == "verifiableCredential"
        else "EnvelopedVerifiablePresentation"
    )
    if (
        not contexts
        or contexts[0] != "https://www.w3.org/ns/credentials/v2"
        or not isinstance(types, (list, str))
        or expected_type not in (types if isinstance(types, list) else [types])
        or not isinstance(identifier, str)
    ):
        raise HTTPException(
            status_code=422, detail={"error": "unsupported_serialization"}
        )

    prefix = "data:application/"
    if not identifier.startswith(prefix) or "," not in identifier:
        raise HTTPException(status_code=422, detail={"error": "invalid_envelope"})
    media_type, encoded_token = identifier[5:].split(",", 1)
    allowed_media_types = (
        {"application/vc+jwt", "application/jwt"}
        if field == "verifiableCredential"
        else {"application/vp+jwt", "application/jwt"}
    )
    try:
        normalized_media_type = unquote_to_bytes(media_type).decode("ascii").lower()
    except UnicodeDecodeError:
        raise HTTPException(
            status_code=422, detail={"error": "invalid_envelope"}
        ) from None
    if normalized_media_type not in allowed_media_types:
        raise HTTPException(
            status_code=422, detail={"error": "unsupported_serialization"}
        )
    try:
        token = unquote_to_bytes(encoded_token).decode("ascii")
    except (UnicodeDecodeError, ValueError):
        raise HTTPException(
            status_code=422, detail={"error": "invalid_envelope"}
        ) from None
    if token.count(".") != 2 or not all(token.split(".")):
        raise HTTPException(status_code=422, detail={"error": "invalid_envelope"})
    return token


async def _evaluate(
    token: str | dict[str, Any],
    options: dict[str, Any],
    request: Request,
    *,
    organization_id: str,
    policy_id: str,
) -> Response:
    _require_organization_authorization(request, organization_id)
    registry = get_registry()
    service_url = registry.get_service_url("presentation-policies")
    if not service_url:
        raise HTTPException(
            status_code=503, detail="Presentation policy service unavailable"
        )
    response = await proxy_request(
        request,
        service_url,
        f"/v1/presentation-policies/{policy_id}/evaluate",
        body_override=json.dumps(
            {
                "vp_token": token,
                "nonce": options.get("challenge"),
                "audience": options.get("domain"),
            }
        ).encode("utf-8"),
        inject_headers={"Content-Type": "application/json"},
    )
    if response.status_code >= 400:
        return response
    try:
        evaluation = json.loads(bytes(response.body))
    except (AttributeError, TypeError, ValueError, json.JSONDecodeError):
        raise HTTPException(
            status_code=502, detail="Presentation policy returned an invalid response"
        ) from None
    if not isinstance(evaluation, dict):
        raise HTTPException(
            status_code=502, detail="Presentation policy returned an invalid response"
        )
    if evaluation.get("decision") != "allow" or evaluation.get("result") != "passed":
        return JSONResponse(
            status_code=422,
            content={"verified": False, "errors": ["verification_failed"]},
        )
    return JSONResponse(
        {
            "verified": True,
            "results": evaluation,
            "problemDetails": [],
        }
    )


async def _issue_data_integrity_credential(
    credential: dict[str, Any],
    request: Request,
    *,
    organization_id: str,
    template_id: str,
    issuer_did: str,
) -> dict[str, Any]:
    """Issue through Marty's ordinary OID4VCI service path.

    The VC-API endpoint is only a shape adapter. It still resolves the supplied
    issuer DID to the template's authorized active issuer profile, creates a
    pre-authorized transaction, redeems a token, obtains a fresh nonce, and
    submits a cryptographically valid holder proof.
    """
    _require_organization_authorization(request, organization_id)
    document_issuer = _credential_issuer_id(credential)
    if document_issuer is not None and document_issuer != issuer_did:
        raise HTTPException(
            status_code=422,
            detail={
                "error": "issuer_mismatch",
                "error_description": (
                    "credential issuer must match the issuer_did selected for signing"
                ),
            },
        )
    try:
        body = IssuanceCreate(
            organization_id=organization_id,
            credential_template_id=template_id,
            issuer_did=issuer_did,
            credential_document=json.loads(json.dumps(credential)),
        )
    except ValidationError as exc:
        # This model is constructed inside the VC-API shape adapter, after
        # FastAPI has validated the outer request. Translate production
        # issuance validation failures into the same controlled 422 boundary
        # a direct /v1/issuance request would expose, without logging the
        # submitted credential or leaking an internal traceback.
        raise HTTPException(
            status_code=422,
            detail={
                "error": "invalid_credential",
                "validation_errors": exc.errors(
                    include_url=False,
                    include_context=False,
                    include_input=False,
                ),
            },
        ) from None
    initiated = await create_issuance(body, request)
    if initiated.status_code >= 400:
        raise HTTPException(
            status_code=initiated.status_code,
            detail=bytes(initiated.body)[:300].decode("utf-8", errors="replace"),
        )
    try:
        transaction = json.loads(bytes(initiated.body))
    except (AttributeError, TypeError, ValueError, json.JSONDecodeError):
        raise HTTPException(
            status_code=502,
            detail="Marty general issuance API returned an invalid response",
        ) from None
    pre_auth_code = transaction.get("pre_auth_code")
    if not isinstance(pre_auth_code, str) or not pre_auth_code:
        raise HTTPException(
            status_code=502,
            detail="Marty issuance did not return a pre-authorized code",
        )

    registry = get_registry()
    service_url = registry.get_service_url("issuance")
    if not service_url:
        raise HTTPException(status_code=503, detail="Issuance service unavailable")
    client = get_http_client()

    token_response = await client.post(
        f"{service_url}/v1/issuance/token",
        data={
            "grant_type": "urn:ietf:params:oauth:grant-type:pre-authorized_code",
            "pre-authorized_code": pre_auth_code,
        },
        timeout=30.0,
    )
    if token_response.status_code >= 400:
        raise HTTPException(
            status_code=token_response.status_code, detail=token_response.text[:300]
        )
    access_token = token_response.json().get("access_token")
    if not isinstance(access_token, str) or not access_token:
        raise HTTPException(
            status_code=502, detail="Marty issuance did not return an access token"
        )

    nonce_response = await client.post(
        f"{service_url}/v1/issuance/nonce", json={}, timeout=30.0
    )
    if nonce_response.status_code >= 400:
        raise HTTPException(
            status_code=nonce_response.status_code, detail=nonce_response.text[:300]
        )
    nonce = nonce_response.json().get("c_nonce")
    if not isinstance(nonce, str) or not nonce:
        raise HTTPException(
            status_code=502, detail="Marty issuance did not return a proof nonce"
        )
    issuer_base_url = os.environ.get("ISSUER_BASE_URL", "http://gateway:8000").rstrip(
        "/"
    )
    proof = _create_oid4vci_proof(f"{issuer_base_url}/org/{organization_id}", nonce)

    issued = await client.post(
        f"{service_url}/v1/issuance/credential",
        headers={"Authorization": f"Bearer {access_token}"},
        json={"format": "ldp_vc", "proofs": {"jwt": [proof]}},
        timeout=30.0,
    )
    if issued.status_code >= 400:
        raise HTTPException(status_code=issued.status_code, detail=issued.text[:300])
    response = issued.json()
    credentials = response.get("credentials") if isinstance(response, dict) else None
    result = (
        credentials[0]
        if isinstance(credentials, list) and len(credentials) == 1
        else None
    )
    document = result.get("credential") if isinstance(result, dict) else None
    document_issuer = document.get("issuer") if isinstance(document, dict) else None
    document_issuer_id = (
        document_issuer.get("id")
        if isinstance(document_issuer, dict)
        else document_issuer
    )
    if (
        not isinstance(result, dict)
        or result.get("format") != "ldp_vc"
        or not isinstance(document, dict)
        or document_issuer_id != issuer_did
        or not isinstance(document.get("proof"), dict)
        or document["proof"].get("type") != "DataIntegrityProof"
        or document["proof"].get("cryptosuite") != "eddsa-rdfc-2022"
    ):
        raise HTTPException(
            status_code=502,
            detail="Marty issuance did not return a native Data Integrity credential",
        )
    return document


@router.post("/credentials/verify")
async def verify_credential(
    body: VerifyCredentialRequest,
    request: Request,
    organization_id: str = Query(..., min_length=1),
    presentation_policy_id: str = Query(..., min_length=1),
) -> Response:
    """Verify one credential through an active organization policy."""
    return await _evaluate(
        _token_or_unsupported(body.verifiableCredential, "verifiableCredential"),
        body.options,
        request,
        organization_id=organization_id,
        policy_id=presentation_policy_id,
    )


@router.post("/presentations/verify")
async def verify_presentation(
    body: VerifyPresentationRequest,
    request: Request,
    organization_id: str = Query(..., min_length=1),
    presentation_policy_id: str = Query(..., min_length=1),
) -> Response:
    """Verify one presentation through an active organization policy."""
    return await _evaluate(
        _token_or_unsupported(body.verifiablePresentation, "verifiablePresentation"),
        body.options,
        request,
        organization_id=organization_id,
        policy_id=presentation_policy_id,
    )


@router.post("/credentials/issue")
async def issue_credential(
    body: IssueCredentialRequest,
    request: Request,
    organization_id: str = Query(..., min_length=1),
    credential_template_id: str = Query(..., min_length=1),
    issuer_did: str = Query(..., pattern=r"^did:[a-z0-9]+:.+"),
) -> Response:
    """Issue through the normal DID-first, profile-mediated OID4VCI flow."""
    credential = await _issue_data_integrity_credential(
        body.credential,
        request,
        organization_id=organization_id,
        template_id=credential_template_id,
        issuer_did=issuer_did,
    )
    return JSONResponse({"verifiableCredential": credential})

File: rust/services/gateway/src/vc_api.rs
//! Deterministic W3C VC-API compatibility transformations.
//!
//! Cryptographic verification and issuance remain in their canonical Rust
//! services. These functions only adapt public VC-API representations to and
//! from Marty's ordinary service contracts.

use percent_encoding::percent_decode_str;
use serde::{Deserialize, Serialize};
use serde_json::{json, Map, Value};
use thiserror::Error;
use url::Url;

const VCDM_V2_CONTEXT: &str = "https://www.w3.org/ns/credentials/v2";
const PRE_AUTH_GRANT: &str = "urn:ietf:params:oauth:grant-type:pre-authorized_code";

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum VerifiableField {
    Credential,
    Presentation,
}

impl VerifiableField {
    fn envelope_type(self) -> &'static str {
        match self {
            Self::Credential => "EnvelopedVerifiableCredential",
            Self::Presentation => "EnvelopedVerifiablePresentation",
        }
    }

    fn media_types(self) -> &'static [&'static str] {
        match self {
            Self::Credential => &["application/vc+jwt", "application/jwt"],
            Self::Presentation => &["application/vp+jwt", "application/jwt"],
        }
    }
}

impl TryFrom<&str> for VerifiableField {
    type Error = VcApiAdapterError;

    fn try_from(value: &str) -> Result<Self, Self::Error> {
        match value {
            "verifiableCredential" => Ok(Self::Credential),
            "verifiablePresentation" => Ok(Self::Presentation),
            _ => Err(VcApiAdapterError::UnsupportedSerialization),
        }
    }
}

#[derive(Clone, Debug, Eq, Error, PartialEq)]
pub enum VcApiAdapterError {
    #[error("unsupported_serialization")]
    UnsupportedSerialization,
    #[error("invalid_envelope")]
    InvalidEnvelope,
    #[error("invalid_policy_response")]
    InvalidPolicyResponse,
    #[error("invalid_issued_credential")]
    InvalidIssuedCredential,
    #[error("{0}")]
    InvalidOffer(String),
}

impl VcApiAdapterError {
    #[must_use]
    pub fn code(&self) -> &str {
        match self {
            Self::UnsupportedSerialization => "unsupported_serialization",
            Self::InvalidEnvelope => "invalid_envelope",
            Self::InvalidPolicyResponse => "invalid_policy_response",
            Self::InvalidIssuedCredential => "invalid_issued_credential",
            Self::InvalidOffer(message) => message,
        }
    }
}

pub fn adapt_verifiable(value: &Value, field: VerifiableField) -> Result<Value, VcApiAdapterError> {
    if value.as_str().is_some_and(|token| !token.trim().is_empty()) {
        return Ok(value.clone());
    }
    let object = value
        .as_object()
        .ok_or(VcApiAdapterError::UnsupportedSerialization)?;
    if has_data_integrity_proof(object) {
        return Ok(value.clone());
    }
    extract_jose_envelope(object, field).map(Value::String)
}

fn has_data_integrity_proof(object: &Map<String, Value>) -> bool {
    let Some(proof) = object.get("proof") else {
        return false;
    };
    proof
        .as_array()
        .map_or_else(|| vec![proof], |values| values.iter().collect())
        .into_iter()
        .any(|item| {
            item.as_object()
                .and_then(|proof| proof.get("type"))
                .and_then(Value::as_str)
                == Some("DataIntegrityProof")
        })
}

fn extract_jose_envelope(
    object: &Map<String, Value>,
    field: VerifiableField,
) -> Result<String, VcApiAdapterError> {
    let context_valid = match object.get("@context") {
        Some(Value::String(value)) => value == VCDM_V2_CONTEXT,
        Some(Value::Array(values)) => {
            values.first().and_then(Value::as_str) == Some(VCDM_V2_CONTEXT)
        }
        _ => false,
    };
    let type_valid = match object.get("type") {
        Some(Value::String(value)) => value == field.envelope_type(),
        Some(Value::Array(values)) => values
            .iter()
            .any(|value| value.as_str() == Some(field.envelope_type())),
        _ => false,
    };
    let identifier = object.get("id").and_then(Value::as_str);
    if !context_valid || !type_valid || identifier.is_none() {
        return Err(VcApiAdapterError::UnsupportedSerialization);
    }
    let identifier = identifier.expect("checked above");
    let Some((media_type, token)) = identifier
        .strip_prefix("data:")
        .and_then(|value| value.split_once(','))
    else {
        return Err(VcApiAdapterError::InvalidEnvelope);
    };
    let media_type = percent_decode_ascii(media_type)?.to_ascii_lowercase();
    if !field.media_types().contains(&media_type.as_str()) {
        return Err(VcApiAdapterError::UnsupportedSerialization);
    }
    let token = percent_decode_ascii(token)?;
    if token.split('.').count() != 3 || token.split('.').any(str::is_empty) {
        return Err(VcApiAdapterError::InvalidEnvelope);
    }
    Ok(token)
}

fn percent_decode_ascii(value: &str) -> Result<String, VcApiAdapterError> {
    let decoded = percent_decode_str(value)
        .decode_utf8()
        .map_err(|_| VcApiAdapterError::InvalidEnvelope)?;
    decoded
        .is_ascii()
        .then(|| decoded.into_owned())
        .ok_or(VcApiAdapterError::InvalidEnvelope)
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
pub struct OfferSelection {
    pub credential_configuration_id: String,
    pub pre_authorized_code: String,
}

pub fn parse_inline_credential_offer(
    offer_uri: &str,
    expected_issuer: &str,
) -> Result<OfferSelection, VcApiAdapterError> {
    if offer_uri.is_empty() {
        return Err(invalid_offer("credential_offer_uri is missing"));
    }
    let url = Url::parse(offer_uri)
        .map_err(|_| invalid_offer("credential_offer_uri contains invalid JSON"))?;
    let offers = url
        .query_pairs()
        .filter(|(name, _)| name == "credential_offer")
        .map(|(_, value)| value.into_owned())
        .collect::<Vec<_>>();
    if offers.len() != 1 {
        return Err(invalid_offer(
            "credential_offer_uri must contain one inline credential_offer",
        ));
    }
    let offer: Value = serde_json::from_str(&offers[0])
        .map_err(|_| invalid_offer("credential_offer_uri contains invalid JSON"))?;
    let object = offer.as_object().ok_or_else(|| {
        invalid_offer("credential offer issuer does not match the selected organization")
    })?;
    if object.get("credential_issuer").and_then(Value::as_str) != Some(expected_issuer) {
        return Err(invalid_offer(
            "credential offer issuer does not match the selected organization",
        ));
    }
    let configurations = object
        .get("credential_configuration_ids")
        .and_then(Value::as_array)
        .filter(|values| values.len() == 1)
        .and_then(|values| values[0].as_str())
        .filter(|value| !value.is_empty())
        .ok_or_else(|| invalid_offer("credential offer must select exactly one configuration"))?;
    let code = object
        .get("grants")
        .and_then(Value::as_object)
        .and_then(|grants| grants.get(PRE_AUTH_GRANT))
        .and_then(Value::as_object)
        .and_then(|grant| grant.get("pre-authorized_code"))
        .and_then(Value::as_str)
        .filter(|value| !value.is_empty())
        .ok_or_else(|| invalid_offer("credential offer has no pre-authorized code"))?;
    Ok(OfferSelection {
        credential_configuration_id: configurations.into(),
        pre_authorized_code: code.into(),
    })
}

fn invalid_offer(message: &str) -> VcApiAdapterError {
    VcApiAdapterError::InvalidOffer(message.into())
}

#[must_use]
pub fn credential_issuer_id(credential: &Value) -> Option<&str> {
    let issuer = credential.as_object()?.get("issuer")?;
    issuer
        .as_str()
        .or_else(|| issuer.as_object()?.get("id")?.as_str())
}

#[must_use]
pub fn evaluation_request(token: Value, options: &Value) -> Value {
    json!({
        "vp_token": token,
        "nonce": options.get("challenge").cloned().unwrap_or(Value::Null),
        "audience": options.get("domain").cloned().unwrap_or(Value::Null)
    })
}

#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
pub struct AdaptedResponse {
    pub status: u16,
    pub body: Value,
}

pub fn adapt_evaluation(evaluation: &Value) -> Result<AdaptedResponse, VcApiAdapterError> {
    let object = evaluation
        .as_object()
        .ok_or(VcApiAdapterError::InvalidPolicyResponse)?;
    if object.get("decision").and_then(Value::as_str) == Some("allow")
        && object.get("result").and_then(Value::as_str) == Some("passed")
    {
        Ok(AdaptedResponse {
            status: 200,
            body: json!({"verified": true, "results": evaluation, "problemDetails": []}),
        })
    } else {
        Ok(AdaptedResponse {
            status: 422,
            body: json!({"verified": false, "errors": ["verification_failed"]}),
        })
    }
}

pub fn issued_data_integrity_credential<'a>(
    response: &'a Value,
    issuer_did: &str,
) -> Result<&'a Value, VcApiAdapterError> {
    let credentials = response
        .get("credentials")
        .and_then(Value::as_array)
        .filter(|values| values.len() == 1)
        .ok_or(VcApiAdapterError::InvalidIssuedCredential)?;
    let result = credentials[0]
        .as_object()
        .ok_or(VcApiAdapterError::InvalidIssuedCredential)?;
    if result.get("format").and_then(Value::as_str) != Some("ldp_vc") {
        return Err(VcApiAdapterError::InvalidIssuedCredential);
    }
    let document = result
        .get("credential")
        .filter(|value| value.is_object())
        .ok_or(VcApiAdapterError::InvalidIssuedCredential)?;
    let proof = document
        .get("proof")
        .and_then(Value::as_object)
        .ok_or(VcApiAdapterError::InvalidIssuedCredential)?;
    if credential_issuer_id(document) != Some(issuer_did)
        || proof.get("type").and_then(Value::as_str) != Some("DataIntegrityProof")
        || proof.get("cryptosuite").and_then(Value::as_str) != Some("eddsa-rdfc-2022")
    {
        return Err(VcApiAdapterError::InvalidIssuedCredential);
    }
    Ok(document)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[derive(Deserialize)]
    struct Contract {
        schema_version: u32,
        serialization: Vec<SerializationCase>,
        offers: Vec<OfferCase>,
        evaluations: Vec<EvaluationCase>,
        issued_credentials: Vec<IssuedCredentialCase>,
    }

    #[derive(Deserialize)]
    struct SerializationCase {
        field: String,
        input: Value,
        expected: Option<Value>,
        error: Option<String>,
    }

    #[derive(Deserialize)]
    struct OfferCase {
        uri: String,
        expected_issuer: String,
        configuration_id: Option<String>,
        pre_authorized_code: Option<String>,
        error: Option<String>,
    }

    #[derive(Deserialize)]
    struct EvaluationCase {
        input: Value,
        status: u16,
        output: Option<Value>,
        error: Option<String>,
    }

    #[derive(Deserialize)]
    struct IssuedCredentialCase {
        issuer_did: String,
        input: Value,
        expected_issuer: Option<String>,
        error: Option<String>,
    }

    fn contract() -> Contract {
        serde_json::from_str(include_str!(
            "../../../../contracts/vc-api-adapter-behavior.json"
        ))
        .expect("valid VC-API contract")
    }

    #[test]
    fn language_neutral_vc_api_adapter_contract() {
        let contract = contract();
        assert_eq!(contract.schema_version, 1);
        for case in contract.serialization {
            let result = adapt_verifiable(
                &case.input,
                VerifiableField::try_from(case.field.as_str()).expect("field"),
            );
            match case.error {
                Some(error) => assert_eq!(result.expect_err("error").code(), error),
                None => assert_eq!(result.expect("adapted"), case.expected.expect("expected")),
            }
        }
        for case in contract.offers {
            let result = parse_inline_credential_offer(&case.uri, &case.expected_issuer);
            match case.error {
                Some(error) => assert_eq!(result.expect_err("error").code(), error),
                None => {
                    let result = result.expect("offer");
                    assert_eq!(
                        Some(result.credential_configuration_id),
                        case.configuration_id
                    );
                    assert_eq!(Some(result.pre_authorized_code), case.pre_authorized_code);
                }
            }
        }
        for case in contract.evaluations {
            let result = adapt_evaluation(&case.input);
            match case.error {
                Some(error) => assert_eq!(result.expect_err("error").code(), error),
                None => {
                    let result = result.expect("evaluation");
                    assert_eq!(result.status, case.status);
                    assert_eq!(Some(result.body), case.output);
                }
            }
        }
        for case in contract.issued_credentials {
            let result = issued_data_integrity_credential(&case.input, &case.issuer_did);
            match case.error {
                Some(error) => assert_eq!(result.expect_err("error").code(), error),
                None => assert_eq!(
                    result
                        .expect("credential")
                        .get("issuer")
                        .and_then(Value::as_str),
                    case.expected_issuer.as_deref()
                ),
            }
        }
    }

    #[test]
    fn issuer_identifier_accepts_both_vcdm_shapes() {
        assert_eq!(
            credential_issuer_id(&json!({"issuer": "did:example:1"})),
            Some("did:example:1")
        );
        assert_eq!(
            credential_issuer_id(&json!({"issuer": {"id": "did:example:2"}})),
            Some("did:example:2")
        );
        assert_eq!(credential_issuer_id(&json!({"issuer": 42})), None);
    }
}
